sdonboarding.client/src/components/UITemplate/EditDataTemplate.jsx

```diff
@@ -0,0 +1,284 @@
+/**
+ * This is a template file for editing an entity.
+ * Used by Customer, Product, Store, Sale Pages.
+ * 'type' should be passed to this component for it to fetch the required information.
+ */
+
+import { useCallback, useEffect, useState } from 'react';
+import { PropTypes } from 'prop-types';
+import { useGetEditObject, useSendEditRequest, useEditCancel } from '../../hooks/EditDataHooks';
+import { genericEditDataMethods } from '../../utils/GenericEditDataMethods';
+import { useGetOtherObjectsData } from '../../hooks/commonHooks';
+import { useDispatch } from 'react-redux';
+import { ObjectTypes } from '../../utils/GenericObjects';
+import genericMethods from '../../utils/GenericMethods';
+import { setError } from '../../redux/slices/commonSlice';
+import { CheckIcon } from '@heroicons/react/24/solid';
+
+const EditDataTemplate = ({ type }) => {
+
+    const dispatch = useDispatch();
+    const editObject = useGetEditObject(type);
+    const { labels, formElementType } = genericEditDataMethods.getEditObjectLabels(type);
+
+    /* custom hooks */
+    const [localObject, setLocalObject] = useState(editObject);
+    const sendEditRequest = useSendEditRequest(localObject, type);
+    const getOtherObjectsData = useGetOtherObjectsData();
+    const editCancel = useEditCancel(type);
+
+    /* Only relevant for Sales functionality*/
+    const [localCustomers, setLocalCustomers] = useState(null);
+    const [localProducts, setLocalProducts] = useState(null);
+    const [localStores, setLocalStores] = useState(null);
+
+    /** Fetches all the objects that we will need for Sale's Add dropdowns
+     * currently relevant for Sales only
+     */
+    const fetchOthersData = useCallback(async () => {
+        try {
+            if (type === ObjectTypes.Sale) {
+                const { customers, products, stores } = await getOtherObjectsData();
+
+                /*sorting these objects so that default value is at top*/
+                /*this will help in displaying them at top*/
+
+                let defaultObject = customers.find((customer) => customer.id === localObject.customerId);
+                let restObject = customers.filter((customer) => customer.id !== localObject.customerId);
+                setLocalCustomers([defaultObject, ...restObject]);
+
+                defaultObject = products.find((product) => product.id === localObject.productId);
+                restObject = products.filter((product) => product.id !== localObject.productId);
+                setLocalProducts([defaultObject, ...restObject]);
+
+                defaultObject = stores.find((store) => store.id === localObject.storeId);
+                restObject = stores.filter((store) => store.id !== localObject.storeId);
+                setLocalStores([defaultObject, ...restObject]);
```

Comment:
This one seems working. **But** if we are sure that our data are always returning an array. But what if, accidentally, it become different? Let's say, nested objects, string, etc... Will this still work or do we need to set up some validations?

Reply:
Created a separate function and added the array check.